hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/BlockInputStreamFactory.java

```diff
@@ -0,0 +1,34 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.ozone.client.io;
+
+import org.apache.hadoop.hdds.client.BlockID;
+import org.apache.hadoop.hdds.scm.pipeline.Pipeline;
+import org.apache.hadoop.hdds.scm.storage.BlockInputStream;
+import org.apache.hadoop.hdds.security.token.OzoneBlockTokenIdentifier;
+import org.apache.hadoop.security.token.Token;
+
+/**
+ * Interface used to create BlockInputStreams.
+ */
+public interface BlockInputStreamFactory {
```

Comment:
As we discussed on other channels, we may want to avoid confusion coming from the name "factory" as this is pretty much just an interface to define a method that creates (supplies) the stream, this way providing the possibility to inject mock instances from tests.

We might want to use a Supplier<BlockInputStream> in the constructor where currently the factory is provided, and then from tests we can provide a TestBlockInputStream, or a mocked InputStream, while in the production code we can just provide a new BlockInputstream from a lambda expression.

Reply:
Sorry, my bad, I had some assumptions and overlooked a few things here...
@sodonnel after you asked offline for some extra explanation what I am thinking about here, I went and tried out, and realized that even if it is possible to create a functional interface, and use a lambda, you are right, it does not really express anything better...